File: Common/Action/src/main/java/com/greentree/action/ListenerCloser.java
package com.greentree.action;

@FunctionalInterface
public interface ListenerCloser extends AutoCloseable {
	
	static final class Null implements ListenerCloser {
		
		private Null() {
		}
		
		@Override
		public void close() {
		}
		
	}
	
	ListenerCloser EMPTY = new Null();
	
	@Override
	void close();
	
	static ListenerCloser merge(ListenerCloser... closers) {
		return ()-> {
			for(var c : closers)
				c.close();
		};
	}
	
	static ListenerCloser merge(Iterable<? extends ListenerCloser> closers) {
		return ()-> {
			for(var c : closers)
				c.close();
		};
	}
	
}

File: Commons/Action/src/main/java/com/greentree/commons/action/ListenerCloser.kt
package com.greentree.commons.action

fun interface ListenerCloser : AutoCloseable {

	override fun close()

	object Empty : ListenerCloser {

		override fun close() {
		}
	}

	companion object {

		fun merge(vararg closers: ListenerCloser): ListenerCloser {
			return ListenerCloser {
				for(c in closers)
					c.close()
			}
		}

		fun merge(closers: Iterable<ListenerCloser>): ListenerCloser {
			return ListenerCloser {
				for(c in closers)
					c.close()
			}
		}
	}
}
